src/backend/distributed/transaction/relation_access_tracking.c

```diff
@@ -0,0 +1,539 @@
+ * parallel access).
+ *
+ * We keep an integer per relation and use some of the bits to identify the access types
+ * and access modes.
+ *
+ * We store the access types in the first 3 bits:
+ *  - 0th bit is set for SELECT accesses to a relation
+ *  - 1st bit is set for DML accesses to a relation
+ *  - 2nd bit is set for DDL accesses to a relation
+ *
+ * and, access modes in the next 3 bits:
+ *  - 3rd bit is set for PARALLEL SELECT accesses to a relation
+ *  - 4th bit is set for PARALLEL DML accesses to a relation
+ *  - 5th bit is set for PARALLEL DDL accesses to a relation
+ *
+ */
+typedef struct RelationAccessHashKey
+{
+	Oid relationId;
+} RelationAccessHashKey;
+
+typedef struct RelationAccessHashEntry
+{
+	RelationAccessHashKey key;
+
+	int relationAccessMode;
+} RelationAccessHashEntry;
+
+static HTAB *RelationAccessHash;
+
+
+static void AssociateShardAccessWithRelation(uint64 shardId,
+											 ShardPlacementAccessType accessType);
+static void RecordPlacementAccessToCache(Oid relationId,
+										 ShardPlacementAccessType accessType);
+static RelationAccessMode GetRelationAccessMode(Oid relationId,
+												ShardPlacementAccessType accessType);
+static void RecordParallelRelationAccess(Oid relationId, ShardPlacementAccessType
+										 placementAccess);
+static void RecordParallelRelationAccessToCache(Oid relationId,
+												ShardPlacementAccessType placementAccess);
+
+
+/*
+ * Empty RelationAccessHash, without destroying the hash table itself.
+ */
+void
+ResetRelationAccessHash()
+{
+	hash_delete_all(RelationAccessHash);
+}
+
+
+/*
+ * Allocate RelationAccessHash.
+ */
+void
+AllocateRelationAccessHash(void)
+{
+	HASHCTL info;
+	uint32 hashFlags = 0;
+
+	memset(&info, 0, sizeof(info));
+	info.keysize = sizeof(RelationAccessHashKey);
+	info.entrysize = sizeof(RelationAccessHashEntry);
+	info.hash = tag_hash;
+	info.hcxt = ConnectionContext;
+	hashFlags = (HASH_ELEM | HASH_BLOBS | HASH_CONTEXT);
+
+	RelationAccessHash = hash_create("citus connection cache (relationid)",
+									 8, &info, hashFlags);
+}
+
+
+/*
+ * AssociatePlacementAccessWithRelation associates the placement access to the
+ * distributed relation that the placement belongs to.
+ */
+void
+AssociatePlacementAccessWithRelation(ShardPlacement *placement,
+									 ShardPlacementAccessType accessType)
+{
+	uint64 shardId = INVALID_SHARD_ID;
+
+	if (!ShouldRecordRelationAccess())
+	{
+		return;
+	}
+
+	shardId = placement->shardId;
+
+	AssociateShardAccessWithRelation(shardId, accessType);
+}
+
+
+/*
+ * AssociateShardAccessWithRelation associates the shard access to the
+ * distributed relation that the shard belongs to.
+ *
+ * We implemented this function to prevent accessing placement metadata during
+ * recursive calls of the function itself.
+ */
+static void
+AssociateShardAccessWithRelation(uint64 shardId, ShardPlacementAccessType accessType)
+{
+	Oid relationId = RelationIdForShard(shardId);
```

Comment:
now it strikes me that we could even skip the call to ColocatedShardIdInRelation if this function took a relationId instead

Reply:
makes sense, done.